File: slide/fetch-streaming/fetch-stream-processing.js
var query = 'gopher';
fetch('/big-data.csv').then(function(response) { // HL
  var reader = response.body.getReader(); // HL
  var partialRecord = '';
  var decoder = new TextDecoder();

  function search() {
    return reader.read().then(function(result) { // HL
      partialRecord += decoder.decode(result.value || new Uint8Array, { stream: !result.done }); // HL
      // query logic ...
      // Call reader.cancel("No more reading needed."); when result found early. // HL
      if (result.done) {
        throw Error("Could not find value after " + query);
      }
      return search(); // HL
    })
  }

  return search(); // HL
}).then(function(result) {
  console.log("Got the result! It's '" + result + "'");
}).catch(function(err) {
  console.log(err.message);
});

File: 2016/Go in the browser/fetch-streaming/fetch-stream-processing.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
)

// Search start OMIT

func Search(url, query string) (result string, err error) {
	resp, err := http.Get(url)
	if err != nil {
		// handle error
	}
	defer resp.Body.Close() // HL
	r := csv.NewReader(resp.Body) // HL
	for {
		record, err := r.Read()
		if err == io.EOF {
			return "", fmt.Errorf("could not find value after %q", query)
		}
		if err != nil {
			// handle error
		}
		// query logic ...
	}
}

// Search end OMIT

//func main() {
// main start OMIT
result, err := Search("/large.csv", "gopher")
if err != nil {
	// handle error
}
fmt.Printf("Got the result! It's %q\n", result)
// main end OMIT
//}
